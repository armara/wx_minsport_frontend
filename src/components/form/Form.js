import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { Form as FinalForm } from 'react-final-form'
import arrayMutators from 'final-form-arrays'

import 'components/form/Form.scss'

const Form = ({
  validate,
  onSubmit,
  decorators,
  initialValues,
  subscription,
  className,
  children,
  childrenProps,
  ...rest
}) => (
  <FinalForm
    onSubmit={onSubmit}
    validate={validate}
    decorators={decorators}
    initialValues={initialValues}
    mutators={{
      ...arrayMutators,
    }}
  >
    {({ handleSubmit, form }) => {
      return (
        <form
          onSubmit={handleSubmit}
          className={classnames([className, 'base-form'])}
        >
          {typeof children === 'function'
            ? children({ form, initialValues, ...childrenProps, ...rest })
            : children}
        </form>
      )
    }}
  </FinalForm>
)

Form.defaultProps = {
  className: '',
  initialValues: {},
  childrenProps: {},
  subscription: {},
  decorators: [],
  validate: () => {},
}

Form.propTypes = {
  decorators: PropTypes.arrayOf(PropTypes.func),
  subscription: PropTypes.shape({}),
  className: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
  validate: PropTypes.func,
  childrenProps: PropTypes.shape({}),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.func,
  ]).isRequired,
  initialValues: PropTypes.shape({}),
}

export default Form
