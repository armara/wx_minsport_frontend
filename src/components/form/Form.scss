@import "styles/mixins";

.base-form {
  display: flex;
  flex-flow: column;

  h4 {
    @include form-label-text();
  }
}

.form-some-blocks {
  display: flex;

  & > div {
    flex-grow: 1;
  }

  & > div + div {
    margin-left: 10px;
  }
}
