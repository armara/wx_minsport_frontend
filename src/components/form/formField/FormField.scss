@import 'styles/mixins';

.form-field {
  display: flex;
  flex-flow: column;

  & > label,
  & label {
    @include form-label-text();
    display: flex;
    align-items: center;

    & > .form-field_required:after {
      content: '*';
      margin-left: 4px;
      color: $purple-red;
    }
  }

  & > div {
    width: 100%;
    display: flex;
    flex-flow: column;
    flex-wrap: nowrap;

    &.form-field_with-error {
      padding-bottom: 20px;
      position: relative;
    }

    & > p {
      @include ellipsis();
      @include roboto-regular-text($small, 14px, $purple-red);
      width: inherit;
      position: absolute;
      bottom: 0;
    }
  }

  &.form-field_disabled {
    opacity: 0.5;
    cursor: default;
  }
}
