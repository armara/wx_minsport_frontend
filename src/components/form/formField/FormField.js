/* eslint-disable react/no-unused-prop-types, react/jsx-props-no-spreading, react/prop-types */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { Field } from 'react-final-form'

import Input from 'components/input/Input'
import { composeValidators, mustBeNotNull } from 'components/form/validators'
import ToolTipIcon from 'components/toolTip/icon/ToolTipIcon'

import 'components/form/formField/FormField.scss'

const renderError = (items, name) => {
  const { error, modified, touched } = items.meta

  const nestedError = (error || {})[name]
  const errorMessage = nestedError || error
  return (
    errorMessage &&
    (modified || touched) && <p title={errorMessage}>{errorMessage}</p>
  )
}

const getValidators = (validate, isRequired) =>
  isRequired ? composeValidators([mustBeNotNull, validate]) : validate

const defineIfRequired = (isRequired, isDisabled) => {
  if (isDisabled) return false
  return isRequired
}

const getValidatorsIfDefined = (validate, required) => {
  if (typeof validate === 'function') {
    return {
      validate: getValidators(validate, required),
    }
  }

  if (required) {
    return {
      validate: mustBeNotNull,
    }
  }

  return null
}

const getValidateFieldsIfDefined = validateFields => {
  if (validateFields === null) {
    return null
  }

  return {
    validateFields,
  }
}

const defineIfWithError = (validators, required) => {
  return validators !== null || required
}

const defineErrorRenderer = customErrorRenderer => {
  if (typeof customErrorRenderer === 'function') {
    return customErrorRenderer
  }

  return renderError
}

const FormField = props => {
  const {
    label,
    isDisabled,
    className,
    name,
    inputComponent: InputComponent,
    inputProps,
    validateFields,
    validate,
    renderError: customErrorRenderer,
    isRequired,
    accept,
    dataTip,
    ...rest
  } = props

  const required = defineIfRequired(isRequired, isDisabled)
  const validateIfDefined = getValidatorsIfDefined(validate, required)
  const validateFieldsIfDefined = getValidateFieldsIfDefined(validateFields)
  const withError = defineIfWithError(validateIfDefined, required)
  const errorRenderer = defineErrorRenderer(customErrorRenderer)

  return (
    <div
      className={classnames([
        className,
        'form-field',
        isDisabled ? 'form-field_disabled' : '',
      ])}
    >
      {label && (
        <label htmlFor={name}>
          <span
            className={classnames([
              className,
              required ? 'form-field_required' : '',
            ])}
          >
            {label}
          </span>
          {dataTip && <ToolTipIcon dataTip={dataTip} />}
        </label>
      )}
      {InputComponent ? (
        <Field
          id={name}
          name={name}
          {...validateIfDefined}
          {...validateFieldsIfDefined}
        >
          {items => (
            <div className={withError ? 'form-field_with-error' : ''}>
              <InputComponent
                name={name}
                {...inputProps}
                {...items}
                {...rest}
                disabled={isDisabled}
                accept={accept}
              />
              {withError && errorRenderer(items, name)}
            </div>
          )}
        </Field>
      ) : (
        <Field
          id={name}
          name={name}
          {...validateIfDefined}
          {...validateFieldsIfDefined}
        >
          {items => (
            <div className={withError ? 'form-field_with-error' : ''}>
              <Input
                className="form-input"
                {...inputProps}
                {...items.input}
                {...rest}
                disabled={isDisabled}
              />
              {withError && errorRenderer(items, name)}
            </div>
          )}
        </Field>
      )}
    </div>
  )
}

FormField.defaultProps = {
  className: '',
  label: '',
  inputComponent: null,
  isRequired: false,
  isDisabled: false,
  validate: null,
  validateFields: [],
  renderError: null,
  accept: null,
  dataTip: 'Важно',
}

FormField.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  isRequired: PropTypes.bool,
  isDisabled: PropTypes.bool,
  renderError: PropTypes.func,
  inputComponent: PropTypes.oneOfType([
    PropTypes.oneOf([null]),
    PropTypes.element,
    PropTypes.func,
  ]),
  validate: PropTypes.func,
  validateFields: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  accept: PropTypes.string,
  dataTip: PropTypes.string,
}

export default FormField
