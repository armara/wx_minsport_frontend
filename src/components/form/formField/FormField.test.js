import React from 'react'
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'

import Form from 'components/form/Form'
import FormField from 'components/form/formField/FormField'

configure({ adapter: new Adapter() })

describe('<FormField />', () => {
  it('renders correctly default props', () => {
    const tree = renderer
      .create(
        <Form onSubmit={() => {}} validate={() => {}}>
          <FormField
            isRequired
            label="Тип организации"
            id="type"
            name="type"
            inputProps={{
              menuId: 'foundation-type_dropdown',
              items: [{ id: 1, item: 'Все типы организаций' }],
            }}
          />
        </Form>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
