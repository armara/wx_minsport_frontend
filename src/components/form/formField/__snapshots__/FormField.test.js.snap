// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> renders correctly default props 1`] = `
<form
  className="base-form"
  onSubmit={[Function]}
>
  <div
    className="form-field form-field_required"
  >
    <label
      htmlFor="type"
    >
      Тип организации
    </label>
    <div
      className="form-field_with-error"
    >
      <input
        className="form-input base-input input"
        disabled={false}
        id="type"
        items={
          Array [
            Object {
              "id": 1,
              "item": "Все типы организаций",
            },
          ]
        }
        menuId="foundation-type_dropdown"
        name="type"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </div>
</form>
`;
